### PR TITLE
chase-symlinks/systemctl: let's handle cases without /proc/ better

### PR DESCRIPTION
**Pull request by poettering**
Alternative to #25613 #25581

### CODE REVIEW

**Review by yuwata — APPROVED** [review]
Yeah, `xopendirat()` is better in this case. LGTM.

**Comment by keszybz** [comment]
Thanks, the fix looks nice. But the last commit is bad,  because it'd create a warning that would be emitted in too many places. There's just so many places (as previous discussions revealed) where the tools are called without /proc.

**Comment by poettering** [comment]
But yeah, that's the point of them, we should get them fixed.

**Comment by poettering** [comment]
I mean, what are the options:

1. Stop using O_PATH and go back to everything path based.
2. Start warning that we can't work properly if /proc is not mounted, to get the ball rolling
3. Stick the head in the sand, and ignore things, always add all codepaths twice once it pops up.

Which one do you propose?

I am pretty sure O_PATH is the way to go, and we should get the distros to fix their execution environments.

**Comment by yuwata** [comment]
> I mean, what are the options:
> 
>     1. Stop using O_PATH and go back to everything path based.
> 
>     2. Start warning that we can't work properly if /proc is not mounted, to get the ball rolling
> 
>     3. Stick the head in the sand, and ignore things, always add all codepaths twice once it pops up.
> 
> 
> Which one do you propose?
> 
> I am pretty sure O_PATH is the way to go, and we should get the distros to fix their execution environments.

I like the current form, that is, the option 2 is the way to go.
But, the warning added by the last commit should be suppressed with recently added option `--no-warn`.
We can backport only the second commit into the stable branch.

File: src/systemctl/systemctl.c [hunk 1]
```diff
@@ -1148,6 +1149,13 @@ static int run(int argc, char *argv[]) {
         if (r <= 0)
                 goto finish;
 
+        if (proc_mounted() == 0)
```
Comment (yuwata) [inline src/systemctl/systemctl.c:1152]: Please conditionalized with `!arg_no_warn`.

**Review by yuwata — APPROVED** [review]

File: src/systemctl/systemctl.c (hunk 1 above)
Comment (keszybz) [reply src/systemctl/systemctl.c:1152]: OK. So with `arg_no_warn` check this would work for me.

**Comment by poettering** [comment]
But making this subject to arg_no_warn would then also mean this would never pop up in rpm scriplet invocation, since `--no-warn` is specified there.

Which would make the exercise pointless... The goal here after all is to get rpm build env maintainers to start invoking things with /proc/ properly mounted.

**Comment by mrc0mmand** [comment]
> But making this subject to arg_no_warn would then also mean this would never pop up in rpm scriplet invocation, since `--no-warn` is specified there.
> 
> Which would make the exercise pointless... The goal here after all is to get rpm build env maintainers to start invoking things with /proc/ properly mounted.

I mean, we can go with `--no-warn` for now to make Fedora/RHEL/etc. happy again as soon as possible and make a note in the NEWS file that states running `systemctl` and so on without `/proc/` will be deprecated in v254, giving RPM folks and others time to fix it (I know, it's been years already). Then follow it by making the warning unconditional in v253 as another push. Going "nuclear" from the start seems a bit unnecessary.

**Comment by poettering** [comment]
I'd not call a warning "going nuclear"...

**Comment by keszybz** [comment]
> so what's the strategy then to get your chroot environments fixed?

I don't have a strategy. But throwing an error is not a strategy either. Errors should be thrown for things for which we can tell recipient to do something. This patch (w/o `--no-warn`) would immediately generate very highly visible and verbose warnings in a place where end-users see them. I find the approach of degrading the user experience without a clear plan forward not acceptable.

In this case, the problem, arguably, is in rpm or dnf. So we could talk with the rpm or dnf maintainers and ask them to change this. I talked with rpm maintainers before about related issues (how rpm behaves without `/proc`), and their stance is that the caller should set up the environment incl. `/proc`. Thus, maybe dnf maintainers could handle this. But I don't think they have a plan or any spare resources for this, and also people call `rpm` directly, which changes in `dnf` wouldn't cover, and `dnf` is not the only way to call rpm, because there's also zypper and packagekit, etc. So overall, actually getting this fixed would require talking with different parties and convincing them to change their software. There's also the problem that older versions of rpm are still used, and unless we declare that new systemd cannot be used with them, we would need to handle this somehow too.

It isn't enough to be "right". Doing this patch with the user-visible warnings wouldn't actually do much to push rpm authors to do anything, but would only inconvenience users and distro maintainers before the warning is inevitably silenced.

**Comment by mrc0mmand** [comment]
I guess that depends on context, but, for example, doing `dnf group install GNOME` with this PR into a different root yields quite a lot of warnings:

```
# dnf --installroot $PWD/root -y --releasever rawhide --nogpgcheck group install "GNOME" |& tee out.log
...
# grep '/proc/ is not mounted' out.log | wc -l
89
```

and in some cases it's multiple warnings per package:

```
  Running scriptlet: libvirt-daemon-8.10.0-1.fc38.x86_64               937/1136 
⚠️ /proc/ is not mounted. This is not a supported mode of operation. Please fix
your invocation environment to mount /proc/ and /sys/ properly. Proceeding anyway.
Your mileage may vary.
⚠️ /proc/ is not mounted. This is not a supported mode of operation. Please fix
your invocation environment to mount /proc/ and /sys/ properly. Proceeding anyway.
Your mileage may vary.
⚠️ /proc/ is not mounted. This is not a supported mode of operation. Please fix
your invocation environment to mount /proc/ and /sys/ properly. Proceeding anyway.
Your mileage may vary.
Created symlink /etc/systemd/system/sockets.target.wants/virtproxyd.socket → /usr/lib/systemd/system/virtproxyd.socket.
⚠️ /proc/ is not mounted. This is not a supported mode of operation. Please fix
your invocation environment to mount /proc/ and /sys/ properly. Proceeding anyway.
Your mileage may vary.
```

Maybeif this was an one-time warning, it wouldn't be that bad, but that's a matter of opinion. I still think having a grace period without the warning would be the best way to proceed here.

**Comment by poettering** [comment]
so what does rh actually do with tmpfiles? it already aborts with an error if procfs is not mounted. is rh not actually calling tmpfiles from scriptlets anyway? or was that patched out?

**Comment by poettering** [comment]
> I guess that depends on context, but, for example, doing `dnf group install GNOME` with this PR into a different root yields quite a lot of warnings:

Yeah, but it's your own fault: yo should have mounted procfs and sysfs into $PWD/root first.

I kinda sympathize with the rpm/dnf people who say: these dirs should be mounted before rpm/dnf is called. You didn't hence it falls apart.

I mean, I think it would make a ton of sense for rpm to warn in this case at least, but either way I understand their thinking that the build env is responsible not them.

**Comment by keszybz** [comment]
> I am not sure it's really rpm/dnf that is at fault. it's probably more the tools that invoke rpm/dnf, and do not prepare the environment properly. (That's at least how I understood the rpm/dnf maintainers when this was last discussed.) i.e. the folks who do `mkdir -p $rootdir` before invoking `dnf --installroot=` should also do a `mount -m proc $rootdir/proc` and `mount -m sysfs $root/sys` first.

That turns a single-line call into something that requires five lines and additional state if errors are to be caught and propagated. Scripting this is not a very big deal, but for interactive use this would be very annoying. Ideally this would be handled in rpm. Doing it anywhere else just so cumbersome. I.e. it's rpm that handles its `--root` option, it runs as root, and if it were to set up and tear down the environment, this would be very very nice for all callers. If not `rpm`, then `dnf`. But I think `dnf` doesn't do by itself execute anything, so it'd be cleaner for `rpm` which calls programs in the chroot to do everything its own.

> I mean, I think it would make a ton of sense for rpm to warn in this case at least, but either way I understand their thinking that the build env is responsible not them.

I disagree. rpm warning is just pushing responsibility onto somebody else, with the problem that the most reasonable place to fix this is in rpm itself, so there's nobody who would fix responsible for this, and it's quite likely this will not get fixed.

--

But meh, I think enough has been said. I care about getting this patch out in Fedora, so that we can fix the regression. So I'll merge this as is. Hopefully we can add support for `--no-warn` before the release.

**Comment by keszybz** [comment]
Queued up for v252.3.

**Pull request merged**